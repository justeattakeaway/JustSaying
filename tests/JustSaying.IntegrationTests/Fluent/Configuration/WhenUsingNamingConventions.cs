using JustSaying.IntegrationTests.Fluent;
using JustSaying.Messaging.MessageHandling;
using JustSaying.Naming;
using JustSaying.TestingFramework;
using Microsoft.Extensions.DependencyInjection;
using Newtonsoft.Json;

namespace JustSaying.Fluent.Configuration;

public class WhenUsingNamingConventions(ITestOutputHelper outputHelper) : IntegrationTestBase(outputHelper)
{
    private class TestNamingConvention : ITopicNamingConvention, IQueueNamingConvention
    {
        private readonly DefaultNamingConventions _default;

        public TestNamingConvention()
        {
            _default = new DefaultNamingConventions();
        }

        public string TopicName<T>()
        {
            return $"beforetopic-{_default.TopicName<T>()}-aftertopic";
        }

        public string QueueName<T>()
        {
            return $"beforequeue-{_default.QueueName<T>()}-afterqueue";
        }
    }

    [AwsFact]
    public async Task ThenTheNamingConventionIsApplied()
    {
        var services = GivenJustSaying()
            .ConfigureJustSaying(builder => builder.Messaging(m =>
                m.WithQueueNamingConvention<TestNamingConvention>()
                    .WithTopicNamingConvention<TestNamingConvention>()))
            .ConfigureJustSaying((builder) =>
                builder.Publications(pub => pub.WithTopic<SimpleMessage>())
                    .Subscriptions(sub => sub.ForTopic<SimpleMessage>()))
            .AddSingleton<IHandlerAsync<SimpleMessage>, InspectableHandler<SimpleMessage>>()
            .AddSingleton<TestNamingConvention>();

        string json = "";
        await WhenAsync(
            services,
            async (publisher, listener, cancellationToken) =>
            {
                await listener.StartAsync(cancellationToken);
                await publisher.StartAsync(cancellationToken);

                var listenerJson = JsonConvert.SerializeObject(listener.Interrogate(), Formatting.Indented);
                var publisherJson = JsonConvert.SerializeObject(publisher.Interrogate(), Formatting.Indented);

                json = string.Join($"{Environment.NewLine}{Environment.NewLine}",
                        listenerJson,
                        publisherJson)
                    .Replace(UniqueName, "integrationTestQueueName", StringComparison.Ordinal);
            });

        json.ShouldMatchApproved(opt => opt.SubFolder("Approvals"));
    }
}
