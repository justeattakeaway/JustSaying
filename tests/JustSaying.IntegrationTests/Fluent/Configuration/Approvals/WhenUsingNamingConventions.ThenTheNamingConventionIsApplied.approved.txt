{
  "Region": "eu-west-1",
  "Middleware": {
    "Middlewares": [
      {
        "MessageType": "SimpleMessage",
        "QueueName": "beforequeue-simplemessage-afterqueue",
        "MiddlewareChain": [
          "MessageContextAccessorMiddleware",
          "ErrorHandlerMiddleware",
          "LoggingMiddleware",
          "StopwatchMiddleware",
          "SqsPostProcessorMiddleware",
          "HandlerInvocationMiddleware`1[JustSaying.TestingFramework.SimpleMessage]"
        ]
      }
    ]
  },
  "PublishedMessageTypes": {},
  "SubscriptionGroups": {
    "Groups": [
      {
        "Name": "beforequeue-simplemessage-afterqueue",
        "ConcurrencyLimit": 10,
        "Multiplexer": {
          "ChannelCapacity": 100,
          "ReaderCount": 1
        },
        "ReceiveBuffers": [
          {
            "BufferSize": 10,
            "QueueName": "beforequeue-simplemessage-afterqueue",
            "Region": "eu-west-1",
            "Prefetch": 10
          }
        ]
      }
    ]
  }
}

{
  "Region": "eu-west-1",
  "Middleware": {
    "Middlewares": []
  },
  "PublishedMessageTypes": {
    "SimpleMessage": {
      "Arn": "arn:aws:sns:us-east-1:000000000000:beforetopic-simplemessage-aftertopic"
    }
  },
  "SubscriptionGroups": {
    "Groups": []
  }
}